File: 0229-majority-element-ii/0229-majority-element-ii.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        List<Integer> ans = new ArrayList<Integer>();
        HashMap<Integer, Integer> map = new HashMap<>(); //stores only 3 entries
        
        for(int num: nums){
            if(map.containsKey(num)){
                map.put(num, map.get(num)-1);
            }else if(map.size() == 3){ 
                // reduce count by 1 and add element
                
                // reduce count and remove if 0
                /*for(int key : map.keySet()){
                    map.put(key, map.get(key)-1);
                }
                // removing 0 valued keys to maintain hashmap size less than 3
                map.entrySet().removeIf(entry -> entry.getValue() == 0);*/
                
                // create a new temp set to avoid concurrency exception
                for (int k : new HashSet<Integer>(map.keySet()))
                if (map.get(k) == 1)
                    map.remove(k);
                else
                    map.put(k, map.get(k) - 1);
                // add element
                map.put(num, 1);
            }else{ 
               // seen new element, and map size < 3 - add to map
               map.put(num, 1); 
            }
        }
        
        // verify and add to ans
        for(int num : map.keySet()){
            int count = 0;
            for(int i = 0; i < nums.length; i++){
                if(nums[i] == num) count++;
            }
            if(count > nums.length/3)
                ans.add(num);
        }
        return ans;
    }
}

File: 0229-majority-element-ii/0229-majority-element-ii.kt
class Solution {
    fun majorityElement(nums: IntArray): List<Int> {
        // extension of Boyre Moore majority vote algorithm
        
        if(nums.size < 2) return nums.toList();
        
        // var ans = listOf<Int>(); // readonly list of kotlin
        val ans = ArrayList<Int>();
        val map = HashMap<Int,Int>();
        
        for(n in nums){
            if(map.containsKey(n)){
                map.put(n, map.get(n)!!.plus(1));
            }else if(map.size == 3){
                for(k in map.keys){
                    val v = map.get(k)!!;
                    map.put(k, v-1);
                }
                map.entries.removeIf {it.value == 0}
                map.put(n,1);
            }else{
                map.put(n,1);
            }
        }
        
        // validate the expected ans
        for(k in map.keys){
            var count = 0;
            for((i, num) in nums.withIndex()){
                if(nums[i] == k){
                    count++;
                }
            }
            if(count > nums.size/3){
                // ans += k; // can be used for for readonly list
                ans.add(k);
            }
        }
        
        return ans;
    }
}